docker-compose.yml

```diff
@@ -0,0 +1,30 @@
+version: "3"
+
+services:
+  web:
+    build:
+      context: .
+    volumes:
+      - .:/app
+      # don't mount tmp directory (https://github.com/Shopify/bootsnap/issues/177)
+      - /app/tmp
+    ports:
+      - "3000:3000"
+    environment:
+      # https://github.com/Shopify/bootsnap/issues/262
+      ENABLE_BOOTSNAP: 'false'
```

Comment:
I don't see where ENABLE_BOOTSNAP would be used right now. Did you maybe forget to check in your changes to config/boot.rb here?

Reply:
This is me chasing down some Ruby 2.5/2.7 vs. Ubuntu 18/20 weirdness when migrating away from the `ruby:2.7` base image per @gravitystorm’s recommendation above. I might be able to back this out if the latest Ubuntu 20.04 + Ruby 2.7 commit builds cleanly.

Reply:
Some details on why Bootsnap. Without these references in `docker-compose.yml`, we see this output:

```
+ docker-compose run --rm web bundle exec rake db:migrate
Starting openstreetmap-website_db_1 ... done
Errno::ENOENT: No such file or directory - bs_fetch:atomic_write_cache_file:chmod
/var/lib/gems/2.7.0/gems/bootsnap-1.5.1/lib/bootsnap/compile_cache/iseq.rb:30:in `fetch'
/var/lib/gems/2.7.0/gems/bootsnap-1.5.1/lib/bootsnap/compile_cache/iseq.rb:30:in `fetch'
/var/lib/gems/2.7.0/gems/bootsnap-1.5.1/lib/bootsnap/compile_cache/iseq.rb:47:in `load_iseq'
/var/lib/gems/2.7.0/gems/bootsnap-1.5.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `require'
/var/lib/gems/2.7.0/gems/bootsnap-1.5.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `block in require_with_bootsnap_lfi'
/var/lib/gems/2.7.0/gems/bootsnap-1.5.1/lib/bootsnap/load_path_cache/loaded_features_index.rb:92:in `register'
/var/lib/gems/2.7.0/gems/bootsnap-1.5.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require_with_bootsnap_lfi'
/var/lib/gems/2.7.0/gems/bootsnap-1.5.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:31:in `require'
/var/lib/gems/2.7.0/gems/activesupport-6.0.3.4/lib/active_support/dependencies.rb:324:in `block in require'
/var/lib/gems/2.7.0/gems/activesupport-6.0.3.4/lib/active_support/dependencies.rb:291:in `load_dependency'
/var/lib/gems/2.7.0/gems/activesupport-6.0.3.4/lib/active_support/dependencies.rb:324:in `require'
/var/lib/gems/2.7.0/gems/bootstrap-4.5.3/lib/bootstrap.rb:61:in `register_rails_engine'
/var/lib/gems/2.7.0/gems/bootstrap-4.5.3/lib/bootstrap.rb:11:in `load!'
/var/lib/gems/2.7.0/gems/bootstrap-4.5.3/lib/bootstrap.rb:75:in `<main>'
/var/lib/gems/2.7.0/gems/bootsnap-1.5.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `require'
/var/lib/gems/2.7.0/gems/bootsnap-1.5.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `block in require_with_bootsnap_lfi'
/var/lib/gems/2.7.0/gems/bootsnap-1.5.1/lib/bootsnap/load_path_cache/loaded_features_index.rb:92:in `register'
/var/lib/gems/2.7.0/gems/bootsnap-1.5.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require_with_bootsnap_lfi'
/var/lib/gems/2.7.0/gems/bootsnap-1.5.1/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:31:in `require'
/var/lib/gems/2.7.0/gems/bundler-2.2.3/lib/bundler/runtime.rb:66:in `block (2 levels) in require'
/var/lib/gems/2.7.0/gems/bundler-2.2.3/lib/bundler/runtime.rb:61:in `each'
/var/lib/gems/2.7.0/gems/bundler-2.2.3/lib/bundler/runtime.rb:61:in `block in require'
/var/lib/gems/2.7.0/gems/bundler-2.2.3/lib/bundler/runtime.rb:50:in `each'
/var/lib/gems/2.7.0/gems/bundler-2.2.3/lib/bundler/runtime.rb:50:in `require'
/var/lib/gems/2.7.0/gems/bundler-2.2.3/lib/bundler.rb:174:in `require'
/app/config/application.rb:19:in `<top (required)>'
/app/rakefile:4:in `require_relative'
/app/rakefile:4:in `<top (required)>'
/usr/share/rubygems-integration/all/gems/rake-13.0.1/exe/rake:27:in `<top (required)>'
(See full trace by running task with --trace)
```

With these lines included, `rake db:migrate` runs to completion.

Reply:
Clarification: the error output above happens on my local machine (MacOS, Docker 2.3). Same process seems happy on GH Actions: https://github.com/migurski/openstreetmap-website/runs/1618711125

Reply:
Yeah, on Github, it's pretty much a no-op, as this environment variable isn't being used anywhere b/c config/boot.rb isn't part of this pull request. I also haven't seen this before on Ubuntu 20.04 host.

Reply:
Any advice? I don’t know what Bootsnap is or does, I’m just blindly searching error messages here.

Reply:
In the Github issue you've linked to, they also replaced one line in a file named  `config/boot.rb`:

On my local box, there's one line to load the bootsnap/setup module, which is used to speed up Rails application loading:
```
require "bootsnap/setup" # Speed up boot time by caching expensive operations.
```

They replace that line by making the loading process depend  on your new environment variable ENABLE_BOOTSNAP:

```
require 'bootsnap/setup' if ENV.fetch("ENABLE_BOOTSNAP", "true") == "true"
```

Reply:
That makes sense — I’ve added this check in `config/boot.rb` and it seems to now work both locally and on GH. Thanks!